### PR TITLE
add shuffled deck function

### PR DESCRIPTION
**Pull request by aquino-a**
The changes are:
-add function to get shuffled deck of cards. Blackjack requires the cards to be shuffled before dealing them.  This PR will fulfill that requirement.

### CODE REVIEW

File: core/src/main/java/com/coderanch/blackjack/Cards.java [hunk 1]
```diff
@@ -20,6 +21,8 @@
  */
 final class Cards {
 
+    private static final Random RANDOM = new Random();
```
Comment (nibsi) [inline core/src/main/java/com/coderanch/blackjack/Cards.java]: Don't declare non-constants as static fields. `Random` is a service, so inject it as a constructor parameter into normal classes, or in the case of static utility classes (such as this one) inject it directly into the static method:

```java
static Stack<Card> getShuffledStandardDeck(Random generator) {
```
The reason for this is that it makes unit-testing much easier, because you can inject `Random` implementations that return a known sequence of 'random' values.

File: core/src/main/java/com/coderanch/blackjack/Cards.java [hunk 2]
```diff
@@ -61,4 +64,22 @@ private Cards() {}
     static Set<Card> getStandardDeck() {
         return STANDARD_DECK;
     }
+
+    static Stack<Card> getShuffledStandardDeck(){
+        return getStandardDeck().stream()
+                .map(c -> new CardPair(RANDOM.nextInt(10_000_000), c))
+                .sorted((o1, o2) -> Integer.compare(o1.position, o2.position))
+                .map(cp -> cp.card)
+                .collect(Collectors.toCollection(Stack::new));
```
Comment (nibsi) [inline core/src/main/java/com/coderanch/blackjack/Cards.java]: There is an easier way to shuffle a collection: Convert the collection to a list, call `Collections.shuffle(list, generator)` and then convert the list back to the intended collection type.

File: core/src/main/java/com/coderanch/blackjack/Cards.java [hunk 3]
```diff
@@ -61,4 +64,22 @@ private Cards() {}
...
+
+    private static class CardPair{
+        public final int position;
+        public final Card card;
+
+        private CardPair(int position, Card card) {
+            this.position = position;
+            this.card = card;
+        }
+    }
```
Comment (nibsi) [inline core/src/main/java/com/coderanch/blackjack/Cards.java:86]: I am pleased that you used a new private class to simplify a functional operation, but if you use my suggestion above, you won't need it anymore.

File: core/src/main/java/com/coderanch/blackjack/Cards.java [hunk 4]
```diff
@@ -61,4 +64,22 @@ private Cards() {}
     static Set<Card> getStandardDeck() {
         return STANDARD_DECK;
     }
+
+    static Stack<Card> getShuffledStandardDeck(){
```
Comment (nibsi) [inline core/src/main/java/com/coderanch/blackjack/Cards.java]: Don't use the mostly obsolete `Stack` class from the standard API. If you _really_ want to return a collection that exposes stack operations, return the `Deque` interface instead.

I strongly recommend returning a `Set` instead though, and let the calling method convert it to an `ArrayDeque` if they need stack operations. The reason for this is to keep it consistent with the `getStandardDeck` method, and to ensure that the returned collection contains only distinct cards.

File: core/src/test/java/com/coderanch/blackjack/CardsTest.java [hunk 5]
```diff
@@ -124,4 +124,13 @@ public void getStandardDeck_returnsUnmodifiableSet() {
...
         assertThat(Cards.getStandardDeck().size(), is(equalTo(Rank.values().length * Suit.values().length)));
     }
+
+    /**
+     * Tests that {@link Cards#getShuffledStandardDeck()} returns a set that contains all possible distinct cards that can be
+     * made using a combination of a {@link Rank} and a {@link Suit}.
+     */
+    @Theory
+    public void getShuffledStandardDeck_returnsAllCombinationsOfRankAndSuit() {
+        assertThat(Cards.getShuffledStandardDeck().size(), is(equalTo(Rank.values().length * Suit.values().length)));
```
Comment (nibsi) [inline core/src/test/java/com/coderanch/blackjack/CardsTest.java]: This implementation works for `getStandardDeck_returnsAllCombinationsOfRankAndSuit()` because `getStandardDeck()` returns a `Set`. As long as `getShuffledStandardDeck()` returns a `Stack` instead of a `Set` you can't be sure that it won't contain duplicates, and therefore this test method implementation is incomplete.

The easy way to fix this is to simply return a `Set` from `getShuffledStandardDeck()`.

**Review by nibsi — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/coderanch/blackjack/Cards.java [hunk 6]
```diff
@@ -61,4 +64,22 @@ private Cards() {}
...
+        return getStandardDeck().stream()
+                .map(c -> new CardPair(RANDOM.nextInt(10_000_000), c))
```
Comment (liutaurasvilda) [inline core/src/main/java/com/coderanch/blackjack/Cards.java]: Why the range here is `0` .. `10_000_000` (exclusive)? In particular why `10_000_000`?

File: core/src/main/java/com/coderanch/blackjack/Cards.java [hunk 7]
```diff
@@ -61,4 +64,22 @@ private Cards() {}
...
+                .sorted((o1, o2) -> Integer.compare(o1.position, o2.position))
+                .map(cp -> cp.card)
+                .collect(Collectors.toCollection(Stack::new));
+    }
+
+    private static class CardPair{
```
Comment (liutaurasvilda) [inline core/src/main/java/com/coderanch/blackjack/Cards.java]: Nit: inconsistent formatting around braces.

File: core/src/main/java/com/coderanch/blackjack/Cards.java (hunk 4 above)
Comment (liutaurasvilda) [inline core/src/main/java/com/coderanch/blackjack/Cards.java]: Would be great to see a comment explaining what that standard deck looks like. i.e. N number of cards divided into 4 suits...

File: core/src/main/java/com/coderanch/blackjack/Cards.java [hunk 8]
```diff
@@ -61,4 +64,22 @@ private Cards() {}
...
+        public final int position;
+        public final Card card;
+
+        private CardPair(int position, Card card) {
```
Comment (liutaurasvilda) [inline core/src/main/java/com/coderanch/blackjack/Cards.java]: Nit: consider the sequence of the parameters for readability purposes.

In my brains in particular - _"there is a card at position"_ as opposed to _"there is a position at which is a card"_

File: core/src/main/java/com/coderanch/blackjack/Cards.java (hunk 6 above)
Comment (aquino-a) [reply core/src/main/java/com/coderanch/blackjack/Cards.java]: It's arbitrary.  I didn't want duplicate numbers, so I picked a high number to lessen that outcome.

File: core/src/main/java/com/coderanch/blackjack/Cards.java (hunk 7 above)
Comment (aquino-a) [reply core/src/main/java/com/coderanch/blackjack/Cards.java]: you mean there's no space after 'CardPair'?

File: core/src/main/java/com/coderanch/blackjack/Cards.java (hunk 6 above)
Comment (nibsi) [reply core/src/main/java/com/coderanch/blackjack/Cards.java]: The lesson here is: Never use arbitrary values. If you don't have a good reason to use a specific value, either change your design or make the value configurable.

In this case the issue was resolved by a design change.

File: core/src/main/java/com/coderanch/blackjack/Cards.java [hunk 9]
```diff
@@ -53,12 +54,18 @@ private Cards() {}
     }
 
     /**
-     * Gets a standard deck of cards.
+     * Gets a standard deck of cards. A standard deck contains all combinations of rank and suit
```
Comment (nibsi) [inline core/src/main/java/com/coderanch/blackjack/Cards.java]: Please put each sentence on a new line. Terminate sentences with a period.

Personally I don't think this line adds anything beyond what the `@return` tag already states, and therefore it can be omitted.

I believe Liutauras' comment pertained to the `getShuffledStandardDeck()` lacking any documentation at all.

File: core/src/main/java/com/coderanch/blackjack/Cards.java [hunk 10]
```diff
@@ -53,12 +54,18 @@ private Cards() {}
...
      *
      * @return an unmodifiable set containing all distinct cards that can be made
      *         using a combination of a {@link Rank} and a {@link Suit}.
      */
     static Set<Card> getStandardDeck() {
         return STANDARD_DECK;
     }
+
+    static Set<Card> getShuffledStandardDeck(Random generator) {
```
Comment (nibsi) [inline core/src/main/java/com/coderanch/blackjack/Cards.java:82]: Add documentation to methods.

File: core/src/test/java/com/coderanch/blackjack/CardsTest.java [hunk 11]
```diff
@@ -123,4 +123,13 @@ public void getStandardDeck_returnsUnmodifiableSet() {
...
         assertThat(Cards.getStandardDeck().size(), is(equalTo(Rank.values().length * Suit.values().length)));
     }
+
+    /**
+     * Tests that {@link Cards#getShuffledStandardDeck(Random)} returns a set that contains all possible distinct cards that can be
+     * made using a combination of a {@link Rank} and a {@link Suit}.
+     */
+    @Theory
+    public void getShuffledStandardDeck_returnsAllCombinationsOfRankAndSuit() {
+        assertThat(Cards.getShuffledStandardDeck(new Random()).size(), is(equalTo(Rank.values().length * Suit.values().length)));
```
Comment (nibsi) [inline core/src/test/java/com/coderanch/blackjack/CardsTest.java]: Tests must be deterministic, meaning you can't use random values.

Instead of passing a new `Random`, pass a custom subclass of `Random` that returns a deterministic sequence of values.

Let me know if you need help with this.

**Review by nibsi — CHANGES_REQUESTED** [review]
One of my comments has already been caught by automatic code analysis.

I haven't been able to let automatic code analysis set checks on pull requests yet. After you've pushed an update to your PR and the PR says that Travis has finished checking your code, take a look here to see the results of automatic code analysis on your PR:

https://sonarcloud.io/dashboard?id=com.coderanch%3Ablackjack-parent&pullRequest=15

Click on the number next to "New Code Smells" to see the problems.

File: core/src/main/java/com/coderanch/blackjack/Cards.java [hunk 12]
```diff
@@ -61,4 +61,16 @@ private Cards() {}
     static Set<Card> getStandardDeck() {
         return STANDARD_DECK;
     }
+
+    /**
+     * Gets a shuffled standard deck of cards.
+     *
+     * @param generator Random number generator
+     * @return a shuffled standard deck {@link Cards#getStandardDeck()}.
```
Comment (nibsi) [inline core/src/main/java/com/coderanch/blackjack/Cards.java]: The generated Javadoc will read like this:

> a shuffled standard deck `getStandardDeck()`

This is not a well flowing sentence. It doesn't explain the relation to the `getStandardDeck()` method. It also doesn't specify any properties about the returned value, such as that it contains all possible playing cards and that it is shuffled using a distribution that depends on the random number generator.

Try something like this instead:

```java
* @return an unmodifiable set containing all distinct cards that can be made
*         using a combination of a {@link Rank} and a {@link Suit}.
*         The cards will be shuffled with a distribution that depends on the properties of {@code generator}.
```

File: core/src/main/java/com/coderanch/blackjack/Cards.java [hunk 13]
```diff
@@ -61,4 +61,16 @@ private Cards() {}
     static Set<Card> getStandardDeck() {
         return STANDARD_DECK;
     }
+
+    /**
+     * Gets a shuffled standard deck of cards.
+     *
+     * @param generator Random number generator
```
Comment (nibsi) [inline core/src/main/java/com/coderanch/blackjack/Cards.java]: For both `@param` tags and `@return` tags, the following conventions apply:

- Terminate every sentence with a period.
- Start full sentences with a capital letter.
- Start incomplete sentences with a lower-case letter. Usually the sentence starts with an article (a, an, the).

```java
* @param generator a random number generator that is used to shuffle the cards.
```

File: core/src/main/java/com/coderanch/blackjack/Cards.java [hunk 14]
```diff
@@ -61,4 +61,16 @@ private Cards() {}
...
+     * @return a shuffled standard deck {@link Cards#getStandardDeck()}.
+     */
+    static Set<Card> getShuffledStandardDeck(Random generator) {
+        List<Card> cards = new ArrayList<>(getStandardDeck());
+        Collections.shuffle(cards, generator);
+        return new LinkedHashSet<>(cards);
```
Comment (nibsi) [inline core/src/main/java/com/coderanch/blackjack/Cards.java]: Make the returned set unmodifiable.

File: core/src/test/java/com/coderanch/blackjack/CardsTest.java [hunk 15]
```diff
@@ -123,4 +123,35 @@ public void getStandardDeck_returnsUnmodifiableSet() {
     public void getStandardDeck_returnsAllCombinationsOfRankAndSuit() {
         assertThat(Cards.getStandardDeck().size(), is(equalTo(Rank.values().length * Suit.values().length)));
     }
+
+    /**
+     * Tests that {@link Cards#getShuffledStandardDeck(Random)} returns a set that contains all possible distinct cards that can be
```
Comment (nibsi) [inline core/src/test/java/com/coderanch/blackjack/CardsTest.java]: Keep lines within 120 columns.

File: core/src/test/java/com/coderanch/blackjack/CardsTest.java [hunk 16]
```diff
@@ -123,4 +123,35 @@ public void getStandardDeck_returnsUnmodifiableSet() {
...
+     * made using a combination of a {@link Rank} and a {@link Suit}.
+     */
+    @Theory
+    public void getShuffledStandardDeck_returnsAllCombinationsOfRankAndSuit() {
+        var random = new FixedRandom(List.of(5,4,2,1,3));
+        assertThat("Check that the card count is correct after shuffling", Cards.getShuffledStandardDeck(random).size(), is(equalTo(Rank.values().length * Suit.values().length)));
```
Comment (nibsi) [inline core/src/test/java/com/coderanch/blackjack/CardsTest.java]: Phrase descriptions as an assertion using the word "must".

```java
assertThat(
    "Standard deck must contain expected number of distinct cards.",
    Cards.getShuffledStandardDeck(generator).size(),
    is(equalTo(Rank.values().length * Suit.values().length))
);
```

File: core/src/test/java/com/coderanch/blackjack/CardsTest.java [hunk 17]
```diff
@@ -123,4 +123,35 @@ public void getStandardDeck_returnsUnmodifiableSet() {
...
+    }
+
+    private static class FixedRandom extends Random {
```
Comment (nibsi) [inline core/src/test/java/com/coderanch/blackjack/CardsTest.java:152]: Good job. There are a couple of issues here, but I want to make some general refactors in a separate pull request. This is fine for now.

File: core/src/main/java/com/coderanch/blackjack/Cards.java [hunk 18]
```diff
@@ -61,4 +61,16 @@ private Cards() {}
...
     }
+
+    /**
+     * Gets a shuffled standard deck of cards.
+     *
+     * @param generator Random number generator
+     * @return a shuffled standard deck {@link Cards#getStandardDeck()}.
+     */
+    static Set<Card> getShuffledStandardDeck(Random generator) {
+        List<Card> cards = new ArrayList<>(getStandardDeck());
```
Comment (nibsi) [inline core/src/main/java/com/coderanch/blackjack/Cards.java]: You can use `var` for local variables when their type is clear from context.

**Review by nibsi — CHANGES_REQUESTED** [review]
Well done.

I'm being very nitpicky here, but that's the purpose of this project.

File: core/src/test/java/com/coderanch/blackjack/CardsTest.java (hunk 17 above)
Comment (aquino-a) [reply core/src/test/java/com/coderanch/blackjack/CardsTest.java:152]: thanks

**Review by nibsi — DISMISSED** [review]

File: core/src/test/java/com/coderanch/blackjack/CardsTest.java [hunk 19]
```diff
@@ -123,4 +123,39 @@ public void getStandardDeck_returnsUnmodifiableSet() {
...
+        private int place = 0;
+
+        public FixedRandom(List<Integer> numbers) {
+            this.numbers = numbers;
+        }
+
+        @Override
+        public int nextInt(int bound) {
+            var next = numbers.get(place++) * layer;
+            if(place > numbers.size() - 1){
```
Comment (liutaurasvilda) [inline core/src/test/java/com/coderanch/blackjack/CardsTest.java]: Formatting: please give to formatting the same importance as you give to design or anything else. That not only makes code look neat, but also who look at the source code can perceive that professionals been at work.

Formating issues:
1. After `if` statement there is no space, right before the opening parentheses.
2. There is no space after closing parentheses, right before opening curly brace.

File: core/src/test/java/com/coderanch/blackjack/CardsTest.java [hunk 20]
```diff
@@ -123,4 +123,39 @@ public void getStandardDeck_returnsUnmodifiableSet() {
...
+        assertThat(
+            "Standard deck must contain expected number of distinct cards.",
+            Cards.getShuffledStandardDeck(random).size(),
+            is(equalTo(Rank.values().length * Suit.values().length))
+        );
+    }
+
+    private static class FixedRandom extends Random {
+
+        private List<Integer> numbers;
```
Comment (liutaurasvilda) [inline core/src/test/java/com/coderanch/blackjack/CardsTest.java]: `numbers` can and should be `final`

File: core/src/test/java/com/coderanch/blackjack/CardsTest.java (hunk 19 above)
Comment (liutaurasvilda) [reply core/src/test/java/com/coderanch/blackjack/CardsTest.java]: I normally don't block PRs because of formatting as people tend to fix. However, set auto formatter in your IDE so it complies to standard formatting rules. That way everyone saves time which could be spent on other important aspects of PR.

@nibsi what's your opinion, do you use some formatting tool so it gets reformatted by the build tool? Or you ask developers to take care of their code?

Comment (nibsi) [reply core/src/test/java/com/coderanch/blackjack/CardsTest.java]: @liutaurasvilda I just created a pull request that adds CheckStyle to the Maven build.

Comment (aquino-a) [reply core/src/test/java/com/coderanch/blackjack/CardsTest.java]: sorry, i started to do reformat code in my ide.  I forgot to do it in this pr.

**Review by liutaurasvilda — APPROVED** [review]
Exemplary work and attitude to comments.

**Comment by liutaurasvilda** [comment]
One note: invest a bit more time writing PR description, in the way it is written currently, looks more like a draft.

I'm referring to this:
> adds a function to get a shuffled stack of cards. uses existing getStandardDeck() to get the underlying cards. There is one test checking if all cards are present. 


Perhaps we need to add a template for PR description structure. That would help having descriptions a bit more standardized.

**Review by nibsi — APPROVED** [review]

**Comment by nibsi** [comment]
When you close your PR, please use "Squash and merge" and remove all individual commit messages that GitHub automatically adds to your PR description. Instead, write a new PR description that nicely summarizes the changes you've made to the code base and for what purpose.

**Comment by aquino-a** [comment]
> When you close your PR, please use "Squash and merge" and remove all individual commit messages that GitHub automatically adds to your PR description. Instead, write a new PR description that nicely summarizes the changes you've made to the code base and for what purpose.

Is there something I should do for this PR other than update the description?

**Comment by nibsi** [comment]
@aquino-a when you click "Squash and merge", Github will first show you a new comment field where you can edit the title and description of the PR. You can leave the title as is. Modify the description to make it complete, but short and snappy.

Try it now. We trust you.

**Comment by nibsi** [comment]
Oh yeah, one more thing. After the PR has been merged, you still need to delete the branch. A button to delete your branch will appear right here, in this PR. Use it. :)

**Comment by aquino-a** [comment]
> @aquino-a when you click "Squash and merge"

On this page I only see 'Close with comment'.  Does 'squash and merge' after you select 'Merge pull request'?  I'm not authorized to do that. When I close the PR, it doesn't show me 'Squash and merge', it only gives me an option to delete the branch.

**Comment by nibsi** [comment]
> When I close the PR, it doesn't show me 'Squash and merge'

I suppose only admins or owners can close pull requests.
